### PR TITLE
idea: protobuf serializer

### PR DESCRIPTION
**Pull request by cyx**
## Reasons for changes

In our context, we use protobufs heavily, and probably will continue
to do so for the foreseeable future.

I believe it would be beneficial to have a ProtoSerializer so we can
leverage the other bits that exist here already, but still be able to
utilize protos as your primary models.

## Description

- Add a proto serializer that utilizes the built in proto registry for determining the type name
- Define an `internal/internal.proto` which will be used for the wire format of the proto, similar to `jsonSerializer`.

### CODE REVIEW

File: serializer.go [hunk 1]
```diff
@@ -102,3 +107,59 @@ func NewJSONSerializer(events ...Event) *JSONSerializer {
...
+		return Record{}, errors.New("Unable to marshal non proto event")
+	}
+
+	buf, err := proto.Marshal(pb)
+	if err != nil {
+		return Record{}, err
+	}
+
+	data, err := proto.Marshal(&internal.Envelope{
+		Type: proto.MessageName(pb),
```
Comment (cyx) [inline serializer.go]: Worth noting that protobufs have by definition, their own registry already so I'm choosing to utilize that.

File: serializer.go [hunk 2]
```diff
@@ -102,3 +107,59 @@ func NewJSONSerializer(events ...Event) *JSONSerializer {
...
+		Data: buf,
+	})
+
+	if err != nil {
+		return Record{}, err
```
Comment (cyx) [inline serializer.go]: I'm not quite certain what sort of error pattern to use here, so I'm more leaning on feedback around the error protocol / types / codes to use.

**Comment by savaki** [comment]
I agree that protobuf is a good option for a supported serializer.  To simplify things, I took an initial stab at creating github.com/altairsix/eventsource-protobuf which generates the code for an eventsource.Serializer directly from a .proto file

I've included an example the _examples directory.

**Comment by savaki** [comment]
Closing as you can now use https://github.com/altairsix/eventsource-protobuf to generate Serializers from protobuf definitions.

**Comment by cyx** [comment]
+1

**Pull request closed without merging**